### PR TITLE
Fast power_prune()

### PR DESCRIPTION
**Pull request by ampli**
This PR makes fast-prune() many times faster (on the `And yes [...]` sentence (-short=20) it is **8x** faster).

The idea for this implementation is based on the observation that even though there are no duplicate disjuncts at this stage, there is very much duplication in the left and right sides of the disjuncts (called in the code "disjunct jets" (the discussion about that name is in issue #898).

So the first step of this implementation is to share the memory of these disjunct jets.

Currently, if any connector on one side of a disjunct (named a "disjunct-jet" here) cannot have a match, this disjunct is discarded. With the "shared jets" implementation, when a disjunct-jet is marked with BAD_WORD, all the identical jets are so marked simultaneously and can be discarded without farther detailed checks. Moreover, even `nearest_word` is set simultaneously in all of them, and its next checks in the memory-shared jets are lightweight (skipping them altogether is a FIXME).

The shared disjunct-jets also allow for very efficient connector searches. Currently all the connectors are inserted into the power table, a thing that (in general) causes a very big table, with very long list of connectors in each slot that need to be searched one by one until a match, even though a glance on them shows the that exact duplication among them is high. The disjunct-jet sharing implementation installs connectors only from once from each group of identical disjunct-jets, when their cleanup (when they are marked with BAD_WORD) is done using a refcount that is stored in the suffix_id. (This refcount just mimics the current behavior. (For much greater efficiency I tried to immediately cleanup each bad connector but this leads to less pruning for some reason that I still investigate.)

A side effect of making the disjunct-jet sharing is that the upper bound on the different number of disjuncts for each word is then known. It is then used for a better sizing of the power table arrays.

The disjunct-jet sharing is of course done before power_prune() uses it. Its results are available from the Sentence struct, so in in one-step-parse it is possible to save some overhead on rebuild after the original disjuncts are restored.

BTW, I tried also to use the shared disjuncts-jets idea i to speed up `do_count()` midlle-disjunct comparisons, but no speedup resulted from a reason I still don't know.

While exploring `power_prune()`, I found several more new pruning ideas. One if them (based on nearest shallow connector) I tried to implement in two ways, but the cleaner way yielded much less optimization, which may hint (unless I had a bug) that I need to understand this idea deeper. All of these ideas need a discussion. I also think the expression pruning can be much improved (more pruning), a thing that may save much overhead of power pruning (and maybe of expression pruning itself, but it is very fast already). BTW, this will not directly help the SAT-parser, because I already tested a way to apply the classic power_prune() to expressions, a thing that made the SAT-parser significantly faster (and while writing this I found how to improve that...). I also found a major SAT-parser omission that has a negative performance impact - for some reason it doesn't use the nearest_word idea (but this seems easy to add - not done yet).

BTW, the code still includes a path for the case in which no disjunct-jet sharing is done.
This is used for short sentences, as defined in `api-structures.h` as `SENTENCE_MIN_LENGTH_TRAILING_HASH 10`, which is used for 3 things: jet-sharing, trailing connector encoding, and packing (a better and shorter name is needed...).

Benchmark results:
(12 x(best-of-6)) runs:
 2.660000 -   2.5610 =   -0.0990 +3.9%    data/en/corpus-basic.batch
 8.032000 -   7.5030 =   -0.5290 +7.1%    data/en/corpus-fixes.batch
 2.155000 -   2.1340 =   -0.0210 +1.0%    data/ru/corpus-basic.batch

(5 x(best-of-3)) runs:
26.350000 -  23.6700 =   -2.6800 +11.3%    ../debug/prune/ex/en/and-yet.batch
86.620000 -  78.3400 =   -8.2800 +10.6%    data/en/corpus-fix-long.batch

A language learning benchmark (typically there is smaller speedup):
Dict: dict_50C_2018-12-17_0006 (the dictionary with the 3M line).
Corpus: 9220-0.txt
Speedup: 620.22% (it seems as if it has a bottleneck in `power_prune()`).

### CODE REVIEW

**Comment by ampli** [comment]
"As usual" I find bugs in my PR just after I send them....
This time I tested panic mode and found a memory management problem (heap-buffer-overflow).
I guess adding a test for such cases is desirable.
 
I will force-push a fix, so please don't apply THIS PR for now.

File: link-grammar/api-structures.h [hunk 1]
```diff
@@ -111,6 +117,14 @@ struct word_queue_s
 	word_queue_t *next;
 };
 
+typedef struct
+{
+	Connector **table[2];
+	unsigned int entries[2];
+	unsigned int *num_connectors[2];
+	String_id *csid[2];
+} jet_sharing_t;
```
[the remark this answers is not in the document]
Comment (ampli) [reply link-grammar/api-structures.h]: It is indeed the number of connectors per word, used for power-table sizing.
I will change its name. I'm also try to find out if this struct can be somehow merged with the `power_table` struct. Meanwhile I will change its name.

**Comment by ampli** [comment]
> the code comments do not explain

I will add comments.

**Comment by linas** [comment]
Do you have permissions to merge, when you are ready? I have no objections to having you merge this pull req. I recall that you did not want to do this in the past.

**Comment by ampli** [comment]
> Do you have permissions to merge, when you are ready?

I have the button to do that, so I guess I have the permission.

> I recall that you did not want to do this in the past.

I usually prefer that you review the code and I then fix it according to your comments before it is merged.
But of course I can merge this PR myself (after I fix it, rename variables and change/add comments - this may take me a day).

**Comment by linas** [comment]
Your pick. The work looks good, I have a busy week in front of me; I don't want to block you from other activity that depends on this merge.

**Comment by ampli** [comment]
> I guess adding a test for such cases is desirable.

It turns out there is an appropriate test to detect such a problem:
`ENLangTestCase.test_1_step_parsing_with_long_sentence_with_nulls`
I think this new test was in another branch whose PR (#899) has not been applied yet when I made tests on the current PR.

I have just force-pushed a fixed PR, and my next step is to merge it.

FIXME: The code may need some cleanups, but more thinking is needed on the better implementation:
1. It seems better to merge `jet_sharing_t` with `power_table` and use `alloc_jet_sharing()` similarly to `alloc_count_context()`. In that opportunity a more consistent naming of the involved functions can be done.
2. The logic of the one-step-parse is hard to follow. It is better to rewrite the relevant section in `classic_parse()` to identify this state clearly like:
`if (one_step_parse) { ... }`
3. Since the one-step-parse mode is the expected one to be used from language-bindings (the two step parse is less intuitive) an effort can be done to increase its efficiency by avoiding disjunct copying.

**Comment by ampli** [comment]
> 2\. The logic of the one-step-parse is hard to follow.

I hope that I finally did it right (on PR #908).

> In that opportunity a more consistent naming of the involved functions can be done.

Also `num_cnctrs_per_word` is actually `num_different_jets_per_word`, but I think it can be renamed just to `num_per_word` since the whole struct is about jets.

**Comment by linas** [comment]
>  `num_different_jets_per_word`

change it to `num_per_word; /* Number of different jets per word */`
